google-ads-examples/src/main/java/com/google/ads/googleads/examples/extensions/AddAffiliateLocationExtensions.java

```diff
@@ -0,0 +1,397 @@
+    String matchingFunction =
+        String.format("IN(FeedAttribute[%d, %d], %d)", feedId, attributeIdForChainId, chainId);
+
+    // Creates a CampaignFeed object.
+    CampaignFeed campaignFeed =
+        CampaignFeed.newBuilder()
+            .setFeed(feedResourceName)
+            .addPlaceholderTypes(PlaceholderType.AFFILIATE_LOCATION)
+            .setMatchingFunction(
+                MatchingFunction.newBuilder().setFunctionString(matchingFunction).build())
+            .setCampaign(ResourceNames.campaign(customerId, campaignId))
+            .build();
+
+    // Creates an operation to create the CampaignFeed.
+    CampaignFeedOperation operation =
+        CampaignFeedOperation.newBuilder().setCreate(campaignFeed).build();
+
+    // Adds a CampaignFeed that associates the feed with this campaign for
+    // the AFFILIATE_LOCATION placeholder type.
+    try (CampaignFeedServiceClient feedServiceClient =
+        googleAdsClient.getLatestVersion().createCampaignFeedServiceClient()) {
+      MutateCampaignFeedsResponse response =
+          feedServiceClient.mutateCampaignFeeds(
+              String.valueOf(customerId), ImmutableList.of(operation));
+      System.out.printf(
+          "Campaign feed created with resource name: %s.%n",
+          response.getResultsList().get(0).getResourceName());
+    }
+  }
+
+  /** Gets the feed item attribute ID that specifies the chain ID. */
+  private long getAttributeIdForChainId(FeedMapping feedMapping) {
+    Optional<AttributeFieldMapping> fieldMapping =
+        feedMapping.getAttributeFieldMappingsList().stream()
+            .filter(
+                m -> m.getAffiliateLocationField() == AffiliateLocationPlaceholderField.CHAIN_ID)
+            .findFirst();
+    if (!fieldMapping.isPresent()) {
+      throw new RuntimeException("Affiliate location field mapping isn't setup correctly");
+    }
+    return fieldMapping.get().getFeedAttributeId();
+  }
+
+  /** Removes a list of feeds from the customerID. */
+  private void removeFeeds(GoogleAdsClient googleAdsClient, long customerId, List<Feed> feeds) {
+    List<FeedOperation> removeOperations =
+        feeds.stream()
+            .map(f -> FeedOperation.newBuilder().setRemove(f.getResourceName()).build())
+            .collect(Collectors.toList());
+    try (FeedServiceClient feedServiceClient =
+        googleAdsClient.getLatestVersion().createFeedServiceClient()) {
+      feedServiceClient.mutateFeeds(String.valueOf(customerId), removeOperations);
+    }
+  }
+
+  /** Retrieves all location extension feeds from a specified customer ID. */
+  private List<Feed> getLocationExtensionFeeds(GoogleAdsClient googleAdsClient, long customerId) {
+    String query =
+        "SELECT "
+            + "  customer_feed.resource_name, "
+            + "  customer_feed.feed, "
+            + "  customer_feed.status, "
+            + "  customer_feed.matching_function.function_string "
+            + "FROM "
```

Comment:
Nit: remove extra space after FROM as there is a space before customer_feed. You can also just make this one line.